src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -498,6 +568,157 @@ ShardsMatchingDeleteCriteria(Oid relationId, List *shardIntervalList,
 
 
 /*
+ * SendQueryToShards executes the given query in all placements of the given 
+ * shard list and returns the total affected tuple count. The execution is done
+ * in a distributed transaction and the commit protocol is decided according to
+ * the value of citus.multi_shard_commit_protocol parameter. SendQueryToShards
+ * does not acquire locks for the shards so it is advised to acquire locks to
+ * the shards when necessary.
+ */
+static int
+SendQueryToShards(Query *query, List *shardIntervalList)
+{
+	ListCell *shardIntervalCell = NULL;
+	List *connectionList = NIL;
+	int affectedTupleCount = 0;
+
+	PG_TRY();
+	{
+		foreach(shardIntervalCell, shardIntervalList)
```

Comment:
We should make sure we process the shards in order (e.g. sorted by shard ID), to avoid deadlocks if there are multiple concurrent transactions. It will probably be fine with a single master since all concurrent processes will see the shard placements in the same order, but this will no longer be true in the masterless world.